Signal/src/ViewControllers/ConversationView/ConversationInputTextView.m

```diff
@@ -44,13 +52,89 @@ - (instancetype)init
 
         self.text = nil;
 
-        //        _placeHolder = nil;
-        //        _placeHolderTextColor = [UIColor lightGrayColor];
+        self.placeholderView = [UILabel new];
+        self.placeholderView.text = NSLocalizedString(@"new_message", @"");
+        self.placeholderView.textColor = [UIColor lightGrayColor];
+        self.placeholderView.textAlignment = NSTextAlignmentLeft;
+        [self addSubview:self.placeholderView];
+
+        // We need to do these steps _after_ placeholderView is configured.
+        self.font = [UIFont ows_dynamicTypeBodyFont];
+        self.textContainerInset = UIEdgeInsetsMake(4.0f, 2.0f, 4.0f, 2.0f);
+        self.contentInset = UIEdgeInsetsMake(1.0f, 0.0f, 1.0f, 0.0f);
+
+        [self ensurePlaceholderConstraints];
+        [self updatePlaceholderVisibility];
     }
 
     return self;
 }
 
+- (void)setFont:(UIFont *_Nullable)font
+{
+    [super setFont:font];
+
+    self.placeholderView.font = font;
+}
+
+- (void)setContentInset:(UIEdgeInsets)contentInset
+{
+    [super setContentInset:contentInset];
+
+    [self ensurePlaceholderConstraints];
+}
+
+- (void)setTextContainerInset:(UIEdgeInsets)textContainerInset
+{
+    [super setTextContainerInset:textContainerInset];
+
+    [self ensurePlaceholderConstraints];
+}
+
+- (void)ensurePlaceholderConstraints
+{
+    OWSAssert(self.placeholderView);
+
+    if (self.placeholderConstraints) {
+        [NSLayoutConstraint deactivateConstraints:self.placeholderConstraints];
+    }
+
+    // We align the location of our placeholder with the text content of
+    // this view.  The only safe way to do that is by measuring the
+    // beginning position.
+    UITextRange *beginningTextRange =
+        [self textRangeFromPosition:self.beginningOfDocument toPosition:self.beginningOfDocument];
+    CGRect beginningTextRect = [self firstRectForRange:beginningTextRange];
+
+    CGFloat hInset = beginningTextRect.origin.x;
+    CGFloat topInset = beginningTextRect.origin.y;
+
+    self.placeholderConstraints = @[
+        [self.placeholderView autoPinEdgeToSuperviewEdge:ALEdgeLeft withInset:hInset],
```

Comment:
Do we want to pin to leading for RTL?

Reply:
I don't think that's necessary since placeholderView's text alignment will take care of that for us.